File: shiftclock-app/modules/shiftclock-ble/android/src/main/java/expo/modules/shiftclockble/ShiftclockBleModule.kt
package expo.modules.shiftclockble

import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ShiftclockBleModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ShiftclockBle")

    Events(
      "onDeviceDiscovered",
      "onConnectionStateChanged",
      "onMessageReceived"
    )

    AsyncFunction("scan") {
      throw CodedException(
        "ERR_NOT_IMPLEMENTED",
        "ShiftclockBle.scan is not implemented",
        null
      )
    }

    AsyncFunction("stopScan") {
      throw CodedException(
        "ERR_NOT_IMPLEMENTED",
        "ShiftclockBle.stopScan is not implemented",
        null
      )
    }

    AsyncFunction("connect") { deviceId: String ->
      throw CodedException(
        "ERR_NOT_IMPLEMENTED",
        "ShiftclockBle.connect is not implemented",
        null
      )

      Unit
    }

    AsyncFunction("disconnect") {
      throw CodedException(
        "ERR_NOT_IMPLEMENTED",
        "ShiftclockBle.disconnect is not implemented",
        null
      )
    }

    AsyncFunction("writeAlarm") { payload: Map<String, Any?> ->
      throw CodedException(
        "ERR_NOT_IMPLEMENTED",
        "ShiftclockBle.writeAlarm is not implemented",
        null
      )

      Unit
    }

    AsyncFunction("writeSettings") { payload: Map<String, Any?> ->
      throw CodedException(
        "ERR_NOT_IMPLEMENTED",
        "ShiftclockBle.writeSettings is not implemented",
        null
      )

      Unit
    }
  }
}

File: shiftclock-app/modules/shiftclock-ble/ios/ShiftclockBleModule.swift
import ExpoModulesCore

public final class ShiftclockBleModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ShiftclockBle")

    Events(
      "onDeviceDiscovered",
      "onConnectionStateChanged",
      "onMessageReceived"
    )

    AsyncFunction("scan") { () throws -> Void in
      throw Exception(
        name: "ShiftclockBleNotImplementedException",
        description: "ShiftclockBle.scan is not implemented",
        code: "ERR_NOT_IMPLEMENTED"
      )
    }

    AsyncFunction("stopScan") { () throws -> Void in
      throw Exception(
        name: "ShiftclockBleNotImplementedException",
        description: "ShiftclockBle.stopScan is not implemented",
        code: "ERR_NOT_IMPLEMENTED"
      )
    }

    AsyncFunction("connect") { (_: String) throws -> Void in
      throw Exception(
        name: "ShiftclockBleNotImplementedException",
        description: "ShiftclockBle.connect is not implemented",
        code: "ERR_NOT_IMPLEMENTED"
      )
    }

    AsyncFunction("disconnect") { () throws -> Void in
      throw Exception(
        name: "ShiftclockBleNotImplementedException",
        description: "ShiftclockBle.disconnect is not implemented",
        code: "ERR_NOT_IMPLEMENTED"
      )
    }

    AsyncFunction("writeAlarm") { (_: [String: Any]) throws -> Void in
      throw Exception(
        name: "ShiftclockBleNotImplementedException",
        description: "ShiftclockBle.writeAlarm is not implemented",
        code: "ERR_NOT_IMPLEMENTED"
      )
    }

    AsyncFunction("writeSettings") { (_: [String: Any]) throws -> Void in
      throw Exception(
        name: "ShiftclockBleNotImplementedException",
        description: "ShiftclockBle.writeSettings is not implemented",
        code: "ERR_NOT_IMPLEMENTED"
      )
    }
  }
}
